File: src/main/java/database/finance/BankAccountTable.java
package database.finance;

import com.koenig.commonModel.User;
import com.koenig.commonModel.database.UserService;
import com.koenig.commonModel.finance.Balance;
import com.koenig.commonModel.finance.BankAccount;
import database.NamedParameterStatement;
import database.Table;

import java.sql.Connection;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.ArrayList;
import java.util.List;

public class BankAccountTable extends Table<BankAccount> {
    public static final String NAME = "bankaccount_table";
    private static final String BANK = "bank";
    private static final String BALANCES = "balances";
    private static final String OWNERS = "owners";
    private UserService userService;

    public BankAccountTable(Connection connection, UserService userService) {
        super(connection);
        this.userService = userService;
    }

    @Override
    public String getTableName() {
        return NAME;
    }

    @Override
    protected BankAccount getItem(ResultSet rs) throws SQLException {
        String name = rs.getString(COLUMN_NAME);
        String bank = rs.getString(BANK);
        List<Balance> balances = Balance.getBalances(rs.getBytes(BALANCES));
        List<User> owners = getUsers(userService, rs.getString(OWNERS));
        return new BankAccount(name, bank, owners, balances);
    }


    @Override
    protected String getTableSpecificCreateStatement() {
        return ", " + BANK + " TEXT, " + BALANCES + " BLOB, " + OWNERS + " TEXT";
    }

    @Override
    protected void setItem(NamedParameterStatement ps, BankAccount item) throws SQLException {
        ps.setString(BANK, item.getBank());
        ps.setBytes(BALANCES, Balance.listToBytes(item.getBalances()));
        setUsers(item.getOwners(), ps, OWNERS);
    }


    @Override
    protected List<String> getColumnNames() {
        ArrayList<String> columnNames = new ArrayList<>();
        columnNames.add(BANK);
        columnNames.add(BALANCES);
        columnNames.add(OWNERS);
        return columnNames;
    }

}

File: src/main/java/database/finance/BankAccountTable.kt
package database.finance

import com.koenig.commonModel.database.DatabaseItemTable
import com.koenig.commonModel.database.UserService
import com.koenig.commonModel.finance.Balance
import com.koenig.commonModel.finance.BankAccount
import database.ItemTable
import database.NamedParameterStatement
import java.sql.Connection
import java.sql.ResultSet
import java.sql.SQLException
import java.util.*

class BankAccountTable(connection: Connection, private val userService: UserService) : ItemTable<BankAccount>(connection) {

    override val tableName: String
        get() = NAME


    override val tableSpecificCreateStatement: String
        get() = ", $BANK TEXT, $BALANCES BLOB, $OWNERS TEXT"


    override val columnNames: List<String>
        get() {
            val columnNames = ArrayList<String>()
            columnNames.add(BANK)
            columnNames.add(BALANCES)
            columnNames.add(OWNERS)
            return columnNames
        }

    @Throws(SQLException::class)
    override fun getItem(rs: ResultSet): BankAccount {
        val name = rs.getString(DatabaseItemTable.COLUMN_NAME)
        val bank = rs.getString(BANK)
        val balances = Balance.getBalances(rs.getBytes(BALANCES))
        val owners = getUsers(userService, rs.getString(OWNERS))
        return BankAccount(name, bank, owners, balances)
    }

    @Throws(SQLException::class)
    override fun setItem(ps: NamedParameterStatement, item: BankAccount) {
        ps.setString(BANK, item.bank)
        ps.setBytes(BALANCES, Balance.listToBytes(item.balances))
        setUsers(item.owners, ps, OWNERS)
    }

    companion object {
        val NAME = "bankaccount_table"
        private val BANK = "bank"
        private val BALANCES = "balances"
        private val OWNERS = "owners"
    }

}
